crankshaft/compiler.md

```diff
@@ -0,0 +1,281 @@
+  optimized code turn out to be too optimistic
+
+## Optimized Code vs. Inline Caches and Unoptimized Code
+
+[watch](http://youtu.be/VhpdsjBUS3g?t=18m53s)
+[watch](http://youtu.be/UJPdhx5zTaw?t=26m30s)
+
+### Full Compiler
+
+[slide](http://v8-io12.appspot.com/index.html#54)
+
+- generates code for any JavaScript
+- all code starts unoptimized
+- initial (quick) JIT 
+- is not great and knows (almost) nothing about types
+- needed to start executing code ASAP
+- uses Inline Caches (ICs) to refine knowledge about types at runtime
+
+### Inline Caches
+
+[slide](http://v8-io12.appspot.com/index.html#55)
+
+[Inline Caches implemented in JavaScript](http://mrale.ph/blog/2012/06/03/explaining-js-vms-in-js-inline-caches.html)
+
+- gather knowledge about types while program runs
+- **type dependent** code for operations given specific hidden classes as inputs
+- 1. validate type assumptions (are hidden classes as expected)
+- 2. do work
+- change at runtime via backpatching as more types are discovered to generate new ICs
+  [watch](http://youtu.be/UJPdhx5zTaw?t=28m44s) | [slide](http://v8-io12.appspot.com/index.html#56)
+
+Inline Caches alone without optimizing compiler step make huge performance difference (20x speedup).
+
+### Monomorphism vs. Polymorphism
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=31m30s) | [slide](http://v8-io12.appspot.com/index.html#61)
+
+- operations are monomorphic if hidden classes of arguments are **always** same
+- all others are polymorphic at best and megamorphic at worst
+- monomorphic operations are easier optimized
+
+#### Considerations
+
+- prefer monomorphic over polymorphic functions wherever possible
+
+### Optimizing Compiler
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=33m12s) | [slide](http://v8-io12.appspot.com/index.html#65)
+
+- if function executes a lot it becomes **hot**
+- hot function is re-compiled with optimizing compiler
+  - optimistically 
+  - lots of assumptions made from the calls made to that function so far
+  - type information takend from ICs
+  - operations get inlined speculatively using historic information
+  - monomorphic functions/constructors can be inlined entirely
+  - inlining allows even further optimizations
+
+### Deoptimization
+
+[watch](http://youtu.be/UJPdhx5zTaw?t=36m50s) | [slide](http://v8-io12.appspot.com/index.html#78)
+
+- optimizations are speculative and assumptions are made
+- if assumption is violated
+  - function deoptimized
+  - execution resumes in full compiler code
+  - in short term execution slows down
+  - normal to occur
+  - more info about about function collected
+  - *better* optimization attempted
+  - if assumptions are violated again, deoptimized again and start over 
+- too many deoptimizations cause function to be sent to *deoptimization hell*
+  - considered not optimizable and no optimization is **ever** attempted again
+- certain constructs like `try/catch` are considered not optimizable and functions containing it go straight to
```

Comment:
This was true in Crankshaft but it’s no longer true with the new pipeline! Please remove this.

Reply:
Yes, that's why it's in the document inside the `./crankshaft/` folder.
I'm really sorry I didn't make this more clear :(
I assumed reviewers could just follow the links in the PR description.

I added a bold note to the description to avoid this from happening to other reviewers.